backend/src/main/java/fk/prof/backend/model/assignment/WorkAssignmentSchedule.java

```diff
@@ -0,0 +1,211 @@
+package fk.prof.backend.model.assignment;
+
+import com.google.common.base.Preconditions;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import recording.Recorder;
+
+import java.time.Clock;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.HashSet;
+import java.util.PriorityQueue;
+import java.util.Set;
+import java.util.concurrent.PriorityBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class WorkAssignmentSchedule {
+  private final static Logger logger = LoggerFactory.getLogger(WorkAssignmentSchedule.class);
+  private final static long NANOS_IN_SEC = (long)Math.pow(10, 9);
+
+  private final long referenceTimeInNanos;
+  private final int minAcceptableDelayForWorkAssignmentInSecs;
+  private final int maxAcceptableDelayForWorkAssignmentInSecs;
+  private final int maxConcurrentlyScheduledEntries;
+
+  private final PriorityQueue<ScheduleEntry> entries = new PriorityQueue<>();
+  private Set<RecorderIdentifier> assignedRecorders = new HashSet<>();
+  private final ReentrantLock entriesLock = new ReentrantLock();
+
+  public WorkAssignmentSchedule(int windowDurationInMins,
+                                int windowEndToleranceInSecs,
+                                int schedulingBufferInSecs,
+                                int minAcceptableDelayForWorkAssignmentInSecs,
+                                int maxAcceptableDelayForWorkAssignmentInSecs,
+                                Recorder.WorkAssignment.Builder[] workAssignmentBuilders,
+                                int coveragePct,
+                                int maxConcurrentSlotsAllowed,
+                                int profileDurationInSecs)
+      throws IllegalArgumentException {
+
+    // breathing space at the start, this will usually be a lower value that window tolerance
+    int windowStartToleranceInSecs = schedulingBufferInSecs * 2;
+    //actual time span for which schedule is calculated
+    int effectiveWindowDurationInSecs = (windowDurationInMins * 60) - windowEndToleranceInSecs - windowStartToleranceInSecs;
+    int effectiveProfileDurationInSecs = profileDurationInSecs + schedulingBufferInSecs;
+    int maxScheduleEntriesWithNoOverlap = effectiveWindowDurationInSecs / effectiveProfileDurationInSecs;
```

Comment:
No, its not only about verbosity. In code that manipulates and checks variables heavily (calculations involving loops and conditions etc), it is hard to keep track of things unless they are named just right.

Naming is actually very hard. But don't give up, try to do better.
----
There are only two hard things in Computer Science: cache invalidation and naming things.

-- Phil Karlton